sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -572,25 +611,41 @@ class reduction_impl : private reduction_impl_base {
       typename _T, typename AllocatorT,
       enable_if_t<!IsKnownIdentityOp<_T, BinaryOperation>::value> * = nullptr>
   reduction_impl(buffer<_T, 1, AllocatorT> Buffer, handler &CGH,
-                 const T &Identity, BinaryOperation BOp)
-      : MAcc(std::make_shared<accessor_type>(Buffer)), MIdentity(Identity),
-        MBinaryOp(BOp) {
+                 const T &Identity, BinaryOperation BOp,
+                 bool InitializeToIdentity)
+      : MRWAcc(std::make_shared<rw_accessor_type>(Buffer)), MIdentity(Identity),
+        MBinaryOp(BOp), InitializeToIdentity(InitializeToIdentity) {
     associateWithHandler(CGH);
     if (Buffer.get_count() != 1)
       throw runtime_error("Reduction variable must be a scalar.",
                           PI_INVALID_VALUE);
   }
 
   /// Constructs reduction_impl when the identity value is unknown.
-  // Note that aliasing constructor was used to initialize MAcc to avoid
+  // Note that aliasing constructor was used to initialize MRWAcc to avoid
   // destruction of the object referenced by the parameter Acc.
   template <
-      typename _T = T, class _BinaryOperation = BinaryOperation,
-      enable_if_t<!IsKnownIdentityOp<_T, _BinaryOperation>::value> * = nullptr>
-  reduction_impl(accessor_type &Acc, const T &Identity, BinaryOperation BOp)
-      : MAcc(shared_ptr_class<accessor_type>(shared_ptr_class<accessor_type>{},
-                                             &Acc)),
-        MIdentity(Identity), MBinaryOp(BOp) {
+      typename _T = T,
+      enable_if_t<!IsKnownIdentityOp<_T, BinaryOperation>::value> * = nullptr>
+  reduction_impl(rw_accessor_type &Acc, const T &Identity, BinaryOperation BOp)
+      : MRWAcc(shared_ptr_class<rw_accessor_type>(
+            shared_ptr_class<rw_accessor_type>{}, &Acc)),
+        MIdentity(Identity), MBinaryOp(BOp), InitializeToIdentity(false) {
+    if (Acc.get_count() != 1)
+      throw runtime_error("Reduction variable must be a scalar.",
```

Comment:
I used the word 'scalar' here as opposite to 'vector'. Accessor may be 1-dimensional, but it must not be a true 1-dimensional array, it may have only 1 element, i.e. be a scalar.